File: app/src/main/java/at/shockbytes/corey/ui/activity/WorkoutDetailActivity.java
package at.shockbytes.corey.ui.activity;

import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.ColorDrawable;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.transition.Fade;
import android.util.TypedValue;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.Window;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;

import at.shockbytes.corey.R;
import at.shockbytes.corey.adapter.ExerciseAdapter;
import at.shockbytes.corey.common.core.workout.model.Exercise;
import at.shockbytes.corey.common.core.workout.model.Workout;
import at.shockbytes.util.AppUtils;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * @author Martin Macheiner
 *         Date: 01.11.2015.
 */
public class WorkoutDetailActivity extends AppCompatActivity {

    private static final String ARG_WORKOUT = "arg_workout";

    public static Intent newIntent(Context context, @NonNull Workout workout) {
        return new Intent(context, WorkoutDetailActivity.class)
                .putExtra(ARG_WORKOUT, workout);
    }

    @BindView(R.id.activity_training_detail_imgview_ext_toolbar)
    protected ImageView imgViewExtToolbar;

    @BindView(R.id.activity_training_detail_imgview_body_region)
    protected ImageView imgViewMuscles;

    @BindView(R.id.activity_training_detail_txt_duration)
    protected TextView txtDuration;

    @BindView(R.id.activity_training_detail_txt_exercise_count)
    protected TextView txtExerciseCount;

    @BindView(R.id.activity_training_recyclerview)
    protected RecyclerView recyclerViewExercises;

    @BindView(R.id.activity_training_btn_start)
    protected Button btnStart;

    @BindView(R.id.activity_training_detail_txt_title)
    protected TextView txtName;

    private Workout workout;

    private Unbinder unbinder;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            getWindow().requestFeature(Window.FEATURE_ACTIVITY_TRANSITIONS);
            getWindow().setExitTransition(new Fade(Fade.OUT));
            //getWindow().setEnterTransition(new Explode());
        }
        setContentView(R.layout.activity_workout_detail);
        unbinder = ButterKnife.bind(this);

        if (getIntent().getExtras() != null && getSupportActionBar() != null) {

            getSupportActionBar().setTitle("");
            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);

            workout = getIntent().getParcelableExtra(ARG_WORKOUT);
            if (workout != null) {

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    getWindow().setStatusBarColor(ContextCompat
                            .getColor(this, workout.getDarkColorResForIntensity()));
                }
                getSupportActionBar().setBackgroundDrawable(new ColorDrawable(ContextCompat
                        .getColor(this, workout.getColorResForIntensity())));
            }
        }

        initializeViews();
    }

    private void initializeViews() {

        btnStart.setText(getString(R.string.activity_workout_detail_btn_start, workout.getDisplayableName()));

        imgViewExtToolbar.setBackgroundResource(workout.getColorResForIntensity());
        imgViewMuscles.setImageDrawable(AppUtils.INSTANCE.createRoundedBitmapFromResource(this,
                workout.getImageResForBodyRegion(), workout.getColorResForIntensity()));

        txtName.setText(workout.getDisplayableName());
        txtDuration.setText(getString(R.string.duration_with_minutes, workout.getDuration()));
        txtExerciseCount.setText(getString(R.string.exercises_with_count, workout.getExerciseCount()));

        ExerciseAdapter exerciseAdapter = new ExerciseAdapter(this, new ArrayList<Exercise>());
        recyclerViewExercises.setLayoutManager(new LinearLayoutManager(this));
        recyclerViewExercises.setAdapter(exerciseAdapter);
        exerciseAdapter.setData(workout.getExercises());

        btnStart.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                Intent intent = WorkoutActivity.newIntent(getApplicationContext(), workout);
                ActivityOptionsCompat options = ActivityOptionsCompat
                        .makeSceneTransitionAnimation(WorkoutDetailActivity.this);
                startActivity(intent, options.toBundle());
            }
        });

    }

    @Override
    protected void onStart() {
        super.onStart();
        if (getSupportActionBar() != null) {
            getSupportActionBar().setElevation(0);
        }
    }

    @Override
    protected void onStop() {
        super.onStop();

        int dp = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 8, getResources()
                .getDisplayMetrics());
        if (getSupportActionBar() != null) {
            getSupportActionBar().setElevation(dp);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unbinder.unbind();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        if (item.getItemId() == android.R.id.home) {
            supportFinishAfterTransition();
        }
        return super.onOptionsItemSelected(item);
    }

}

File: app/src/main/java/at/shockbytes/corey/ui/activity/WorkoutDetailActivity.kt
package at.shockbytes.corey.ui.activity

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import androidx.core.app.ActivityOptionsCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import android.transition.Fade
import at.shockbytes.core.ui.activity.base.TintableBackNavigableActivity
import at.shockbytes.corey.R
import at.shockbytes.corey.ui.adapter.ExerciseAdapter
import at.shockbytes.corey.common.core.util.CoreyUtils
import at.shockbytes.corey.common.core.workout.model.Workout
import at.shockbytes.corey.dagger.AppComponent
import at.shockbytes.util.AppUtils
import kotlinx.android.synthetic.main.activity_workout_detail.*

/**
 * Author:  Martin Macheiner
 * Date:    01.11.2015
 */
class WorkoutDetailActivity : TintableBackNavigableActivity<AppComponent>() {

    override val abDefColor: Int = R.color.colorPrimary
    override val abTextDefColor: Int = R.color.white
    override val colorPrimary: Int = R.color.colorPrimary
    override val colorPrimaryDark: Int = R.color.colorPrimaryDark
    override val colorPrimaryText: Int = R.color.colorPrimaryText
    override val sbDefColor: Int = colorPrimaryDark
    override val upIndicator: Int = R.drawable.ic_back_arrow
    override val enableActivityTransition = false

    private lateinit var workout: Workout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.exitTransition = Fade(Fade.OUT)
        setContentView(R.layout.activity_workout_detail)

        workout = intent.getParcelableExtra(ARG_WORKOUT)!!
        setupViews()
    }

    override fun injectToGraph(appComponent: AppComponent?) = Unit

    override fun bindViewModel() = Unit

    override fun unbindViewModel() = Unit

    override fun onStart() {
        super.onStart()
        supportActionBar?.elevation = 0f
    }

    override fun backwardAnimation() {
        super.backwardAnimation()
        activity_training_btn_start.hide()
    }

    override fun onStop() {
        super.onStop()
        supportActionBar?.elevation = AppUtils.convertDpInPixel(8, this).toFloat()
    }

    private fun setupViews() {

        tintSystemBarsWithText(
                actionBarColor = ContextCompat.getColor(this, workout.colorResForIntensity),
                statusBarColor = ContextCompat.getColor(this, workout.darkColorResForIntensity),
                animated = false,
                title = "",
                actionBarTextColor = Color.WHITE,
                useSameColorsForBoth = false
        )

        activity_training_detail_imgview_ext_toolbar.setBackgroundResource(workout.colorResForIntensity)
        activity_training_detail_imgview_body_region.setImageDrawable(AppUtils.createRoundedBitmapFromResource(this,
                workout.imageResForBodyRegion, workout.colorResForIntensity))

        activity_training_detail_txt_title.text = workout.displayableName
        activity_training_detail_txt_duration.text = getString(R.string.duration_with_minutes, workout.duration)
        activity_training_detail_txt_exercise_count.text = getString(R.string.exercises_with_count, workout.exerciseCount)

        activity_training_recyclerview.apply {
            layoutManager = androidx.recyclerview.widget.LinearLayoutManager(this@WorkoutDetailActivity)
            adapter = ExerciseAdapter(this@WorkoutDetailActivity).apply {
                data = workout.exercises
            }
        }

        activity_training_btn_start.setOnClickListener {
            val options = ActivityOptionsCompat.makeSceneTransitionAnimation(this)
            startActivity(WorkoutActivity.newIntent(applicationContext, workout), options.toBundle())
        }

        activity_training_detail_imgview_equipment
                .setImageDrawable(AppUtils.createRoundedBitmapFromResource(this,
                        CoreyUtils.getImageByEquipment(workout.equipment), R.color.equipmentBackground)
                )
    }

    companion object {

        private const val ARG_WORKOUT = "arg_workout"

        fun newIntent(context: Context, workout: Workout): Intent {
            return Intent(context, WorkoutDetailActivity::class.java)
                    .putExtra(ARG_WORKOUT, workout)
        }
    }
}
